### ISSUE TITLE
Completion in context of `if` boolean expressions

### ISSUE THREAD

**Issue by myitcv** [mdempsky/gocode#22]
My question/request is best explained via a couple of examples.

```go
// example 1
package main

type S struct {
    *T
}

func (s *S) DoSomething() {
    // eg 1
    s.
      ^
}

type T int
```

If completion is attempted at the position indicated by `^`, completion successfully returns the full list of candidates:

```
func DoSomething()
var T *T
```

Now consider:

```go
// example 2
package main

type S struct {
    *T
}

func (s *S) DoSomething() {
    // eg 1
    if s.
         ^
}

type T int
```

This time we only get the single candidate:

```
func DoSomething()
```

This appears to be because the declaration of type `T` follows the currently invalid `if` statement; in the case of example 2, this affects parsing sufficiently so as to cause `gocode` to consider the type of the anonymous field (the fact that it's anonymous I don't believe is relevant) declared as `*T` as invalid.

Is there a way to avoid this/improve things? I'm guessing this might be tricky given we end up straying into core `go/parser` territory?

**Comment by mdempsky** [comment]
Thanks for the report. I agree that ideally the second case should include "var T *T" as a suggestion.

I looked into this some, and I couldn't find a good solution off hand. Unfortunately, go/parser and go/types support for ill-formed Go source files is best effort, and this seems like one of the cases where that support fails.

If someone can suggest a non-invasive workaround here, I'm happy to consider it. At the moment though, I've not found anything satisfactory.

**Comment by myitcv** [comment]
Thanks for taking a look in any case @mdempsky

**Comment by brenol** [comment]
Hey @mdempsky,

I took a look at this issue as I wasn't sure what else to tackle and this is what I found out (perhaps you already knew but I am not sure):

If you swap the code to
```go
// example 2
package main

type T int

type S struct {
    *T
}

func (s *S) DoSomething() {
    // eg 1
    if s.
         ^
}
```

Then it properly suggests the correct types as it already knows that T was declared beforehand. I think the problem is in one of these two (perhaps a mashup of them):

a) in the **analyzePackages** method T is marked as Unresolved in the fileAST. (I actually tried to do a fix for this case but I am not sure if this was it)

b) It filters all data up to cursor in the (*tokenIterator).token method, and so T is never returned (as it is positioned after the cursor)

Unfortunately I lack a lot of experience in all packages used here (go/ast; go/types) and I was unable to fix this but I believe it can be fixed if somehow all types can be parsed correctly.
